File: testing/compose/java/chatbot/src/main/java/com/cognizhi/heeczer/teststack/tools/Catalogue.java
package com.cognizhi.heeczer.teststack.tools;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Catalogue {
    private Catalogue() {
    }

    public record ToolTraceEntry(String tool_name, int invoked_at_ms, double output_size) {
    }

    private static final Map<String, Double> OUTPUT_SIZES = Map.of(
            "web_search", 0.0,
            "code_executor", 0.5,
            "document_reader", 0.0,
            "data_analyst", 1.0,
            "plan_reviewer", 0.0,
            "risk_checker", 0.0,
            "summarizer", 0.8,
            "diff_generator", 0.3);

    public static List<ToolTraceEntry> traceForTools(List<String> activeTools) {
        List<ToolTraceEntry> trace = new ArrayList<>();
        for (int index = 0; index < activeTools.size(); index++) {
            String toolName = activeTools.get(index);
            trace.add(new ToolTraceEntry(toolName, index * 25, OUTPUT_SIZES.getOrDefault(toolName, 0.0)));
        }
        return trace;
    }

    public static List<Map<String, Object>> functionSchemas(List<String> activeTools) {
        List<Map<String, Object>> schemas = new ArrayList<>();
        for (String toolName : activeTools) {
            Map<String, Object> input = new LinkedHashMap<>();
            input.put("type", "string");
            Map<String, Object> properties = new LinkedHashMap<>();
            properties.put("input", input);
            Map<String, Object> parameters = new LinkedHashMap<>();
            parameters.put("type", "object");
            parameters.put("properties", properties);
            parameters.put("required", List.of("input"));
            Map<String, Object> function = new LinkedHashMap<>();
            function.put("name", toolName);
            function.put("description", "Synthetic " + toolName + " tool used by the ai-heeczer local stack.");
            function.put("parameters", parameters);
            Map<String, Object> schema = new LinkedHashMap<>();
            schema.put("type", "function");
            schema.put("function", function);
            schemas.add(schema);
        }
        return schemas;
    }
}

File: testing/compose/pydanticai/chatbot/tools/catalogue.py
from __future__ import annotations

from typing import Any, Literal, TypedDict

ToolName = Literal[
    "web_search",
    "code_executor",
    "document_reader",
    "data_analyst",
    "plan_reviewer",
    "risk_checker",
    "summarizer",
    "diff_generator",
]


class ToolContribution(TypedDict):
    tool_call_count: int
    workflow_steps: int
    tokens_prompt: int
    tokens_completion: int
    artifact_count: int
    output_size_proxy: float


TOOL_CONTRIBUTIONS: dict[ToolName, ToolContribution] = {
    "web_search": {"tool_call_count": 1, "workflow_steps": 1, "tokens_prompt": 500, "tokens_completion": 0, "artifact_count": 0, "output_size_proxy": 0.0},
    "code_executor": {"tool_call_count": 1, "workflow_steps": 1, "tokens_prompt": 0, "tokens_completion": 0, "artifact_count": 1, "output_size_proxy": 0.5},
    "document_reader": {"tool_call_count": 1, "workflow_steps": 1, "tokens_prompt": 2000, "tokens_completion": 0, "artifact_count": 0, "output_size_proxy": 0.0},
    "data_analyst": {"tool_call_count": 1, "workflow_steps": 0, "tokens_prompt": 0, "tokens_completion": 800, "artifact_count": 1, "output_size_proxy": 1.0},
    "plan_reviewer": {"tool_call_count": 1, "workflow_steps": 1, "tokens_prompt": 0, "tokens_completion": 0, "artifact_count": 0, "output_size_proxy": 0.0},
    "risk_checker": {"tool_call_count": 1, "workflow_steps": 0, "tokens_prompt": 0, "tokens_completion": 0, "artifact_count": 0, "output_size_proxy": 0.0},
    "summarizer": {"tool_call_count": 1, "workflow_steps": 0, "tokens_prompt": 0, "tokens_completion": 0, "artifact_count": 1, "output_size_proxy": 0.8},
    "diff_generator": {"tool_call_count": 1, "workflow_steps": 0, "tokens_prompt": 0, "tokens_completion": 0, "artifact_count": 1, "output_size_proxy": 0.3},
}


def function_schemas(active_tools: list[ToolName]) -> list[dict[str, object]]:
    return [
        {
            "type": "function",
            "function": {
                "name": tool,
                "description": f"Synthetic {tool} tool used by the ai-heeczer local stack.",
                "parameters": {"type": "object", "properties": {"input": {"type": "string"}}, "required": ["input"]},
            },
        }
        for tool in active_tools
    ]


def trace_for_tools(active_tools: list[ToolName]) -> list[dict[str, object]]:
    return [
        {
            "tool_name": tool,
            "invoked_at_ms": index * 25,
            "output_size": TOOL_CONTRIBUTIONS[tool]["output_size_proxy"],
        }
        for index, tool in enumerate(active_tools)
    ]


def pydantic_ai_tools(active_tools: list[ToolName]) -> list[Any]:
    from pydantic_ai import Tool

    tools: list[Any] = []
    for tool_name in active_tools:
        async def stub(input: str, selected_tool: str = tool_name) -> dict[str, str]:
            return {"tool": selected_tool, "summary": f"synthetic result for {input[:32]}"}

        tools.append(Tool(stub, name=tool_name, description=f"Synthetic {tool_name} local-stack tool"))
    return tools
